core/src/main/scala/kafka/server/ClientQuotaManager.scala

```diff
@@ -310,18 +312,39 @@ class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
    *         rate gets back to the defined quota
    */
   def recordAndGetThrottleTimeMs(session: Session, clientId: String, value: Double, timeMs: Long): Int = {
+    val curTimeMs = Time.SYSTEM.milliseconds()
     val clientSensors = getOrCreateQuotaSensors(session, clientId)
+    val quotaSensor = clientSensors.quotaSensor
     try {
-      clientSensors.quotaSensor.record(value, timeMs, true)
+      quotaSensor.record(value, timeMs, true)
+
+      // Remove the non-violated quota sensor to save memory.
+      quotaViolationLastLoggedTimestamp.remove(quotaSensor.name)
```

Comment:
This is how we expire the records in quotaViolationLastLoggedTimestamp

Reply:
This seems not sufficient for corner cases. (1) In the condition that the quota violation is triggered non consecutively, e.g., in one request, there is violation, and then in the next request, there is no violation. This seems not dealing well. (2) In Kafka rest client, with random client id, we would have no new requests for the client id.

Reply:
Those are valid points. Especially for (2), I think we indeed need a separate thread to expire the entry.